crates/rpc/src/v03/method/simulate_transaction.rs

```diff
@@ -0,0 +1,665 @@
+                dto::DeployAccountTxnTrace {
+                    fee_transfer_invocation: fee.map(map_function_invocation),
+                    validate_invocation: Some(map_function_invocation(val)),
+                    constructor_invocation: Some(map_function_invocation(fun)),
+                },
+            ))
+        }
+        (Some(val), Some(fun), fee)
+            if fun.entry_point_type == Some(dto::EntryPointType::External) =>
+        {
+            Ok(dto::TransactionTrace::Invoke(dto::InvokeTxnTrace {
+                fee_transfer_invocation: fee.map(map_function_invocation),
+                validate_invocation: Some(map_function_invocation(val)),
+                execute_invocation: Some(map_function_invocation(fun)),
+            }))
+        }
+        (Some(val), _, fee) => Ok(dto::TransactionTrace::Declare(dto::DeclareTxnTrace {
+            fee_transfer_invocation: fee.map(map_function_invocation),
+            validate_invocation: Some(map_function_invocation(val)),
+        })),
+        (_, Some(fun), _) => Ok(dto::TransactionTrace::L1Handler(dto::L1HandlerTxnTrace {
+            function_invocation: Some(map_function_invocation(fun)),
+        })),
+        _ => Err(SimulateTransactionError::Internal(anyhow!(
+            "Unmatched transaction trace: '{trace:?}'"
+        ))),
+    }
+}
+
+#[derive(Deserialize, Debug)]
+pub struct SimulateTrasactionInput {
+    block_id: BlockId,
+    transactions: Vec<BroadcastedTransaction>,
+    simulation_flags: dto::SimulationFlags,
+}
+
+#[derive(Debug, Serialize, Eq, PartialEq)]
+pub struct SimulateTransactionResult(pub Vec<dto::SimulatedTransaction>);
+
+#[derive(Debug)]
+pub enum SimulateTransactionError {
+    BlockNotFound,
+    ContractNotFound,
+    ContractError,
+    Internal(anyhow::Error),
+}
+
+impl From<SimulateTransactionError> for RpcError {
+    fn from(value: SimulateTransactionError) -> Self {
+        match value {
+            SimulateTransactionError::BlockNotFound => RpcError::BlockNotFound,
+            SimulateTransactionError::ContractNotFound => RpcError::ContractNotFound,
+            SimulateTransactionError::ContractError => RpcError::ContractError,
+            SimulateTransactionError::Internal(e) => RpcError::Internal(e),
+        }
+    }
+}
+
+impl From<anyhow::Error> for SimulateTransactionError {
+    fn from(err: anyhow::Error) -> Self {
+        Self::Internal(err)
+    }
+}
+
+pub mod dto {
+    use crate::v02::types::reply::FeeEstimate;
+
+    use super::*;
+
+    #[derive(Debug, Deserialize, Serialize, Eq, PartialEq)]
+    pub struct SimulationFlags(pub Vec<SimulationFlag>);
+
+    #[derive(Debug, Deserialize, Serialize, Eq, PartialEq)]
+    pub enum SimulationFlag {
+        #[serde(rename = "SKIP_EXECUTE")]
+        SkipExecute,
+        #[serde(rename = "SKIP_VALIDATE")]
+        SkipValidate,
+    }
+
+    #[derive(Debug, Deserialize, Serialize, Eq, PartialEq)]
+    pub struct Signature(pub Vec<Felt>);
```

Comment:
Shouldn't these `Felt` values be `RpcFelt` instead?

Reply:
Maybe. I don't know how many ways are necessary to dump a hex string :)

Reply:
Ah, to be more precise: RPC types in general use `Felt` in conjunction with `#[serde_as(as = "RpcFelt")]` so that we can have a specific serialization format for RPC responses.

Reply:
I was joking :) Thank you for pointing out, the output DTOs are now using `RpcFelt`.

Reply:
Ah I wrote that comment because I realized that what I wrote was quite different what I meant and very easy to misunderstand.